File: Java/src/weeklyContest240/No3.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

package weeklyContest240;

import java.util.Arrays;
import java.util.Stack;

public class No3 {
    private static final int MOD = 1_000_000_007;

    public int maxSumMinProduct(int[] nums) {
        int n = nums.length;
        int[] t = new int[n + 2];
        for (int i = 1; i <= n; ++i) t[i] = nums[i - 1];
        nums = t;

        int[] left = new int[n], right = new int[n];
        Stack<Integer> stack = new Stack<>();

        // 单调栈的构建过程！
        // 右边第一个小于 nums[i] 的位置
        for (int i = 0; i < n; ++i) {
            if (!stack.isEmpty() && nums[i] < nums[stack.peek()]) right[stack.pop()] = i;
            stack.push(i);
        }
        stack.clear();
        for (int i = n - 1; i >= 0; --i) {
            if (!stack.isEmpty() && nums[i] < nums[stack.peek()]) right[stack.pop()] = i;
            stack.push(i);
        }

        long[] pre = new long[n];
        pre[0] = nums[0];
        for (int i = 1; i < n; ++i) {

        }

        long ans = 0;
        for (int i = 1; i <= n; ++i) {
            ans = Math.max(ans, nums[i] * (pre[right[i]] - pre[left[i]]));
        }
        return (int) ans % MOD;
    }

    public static void main(String[] args) {
        int[] nums = {1, 2, 3, 2};
        No3 no3 = new No3();
        System.out.println(no3.maxSumMinProduct(nums));
    }
}

File: CPP/weekly-contest-272/No3.cc
/**
 * Created by Xiaozhong on 2021/12/19.
 * Copyright (c) 2021/12/19 Xiaozhong. All rights reserved.
 */

#include "vector"
#include "iostream"
#include "algorithm"

using namespace std;

class Solution {
public:
    long long getDescentPeriods(vector<int> &prices) {
        if (prices.size() == 1) return 1;
        long long ans = 0;
        long long len = 1;
        for (int i = 0; i < prices.size(); ++i) {
            if (i == 0) continue;
            if (prices[i - 1] - prices[i] == 1) len++;
            else if (len != 1) {
                ans += (len + 1) * len / 2;
                len = 1;
            } else {
                ans++;
            }
            if (i == prices.size() - 1 && len == 1) ans++;
        }
        if (len != 1) ans += (len + 1) * len / 2;
        return ans;
    }
};

int main() {
    Solution s;
    vector<int> nums = {3, 2, 1, 4};
    cout << s.getDescentPeriods(nums) << endl;
    nums = {12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 4, 3, 10, 9, 8, 7};
    cout << s.getDescentPeriods(nums) << endl;
}